File: capacitor-plugins/capacitor-native/ios/Sources/AonsokuNativePlugin/Image/ImageCacheManager.swift
import Foundation
import GRDB

extension Notification.Name {
    static let aonsokuCoverImageCached = Notification.Name("AonsokuCoverImageCached")
}

enum ImageCacheNotification {
    static let coverArtIdKey = "coverArtId"
}

final class ImageCacheManager {
    private let db: DatabasePool
    private let session: URLSession

    init(db: DatabasePool) {
        self.db = db
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 120
        self.session = URLSession(configuration: config)
    }

    // MARK: - Public API

    func downloadAvatar(username: String, size: String) async throws -> URL {
        guard let credentials = KeychainManager.retrieve() else {
            throw ImageCacheError.noCredentials
        }

        let directory = try ImageCacheUtils.cacheDirectoryURL(createIfNeeded: true)
        let cacheId = ImageCacheUtils.cacheId(for: username)
        // Remove existing file before downloading
        let existingFiles = try FileManager.default.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: nil
        ).filter { $0.lastPathComponent.hasPrefix("\(cacheId).") }
        for file in existingFiles {
            try? FileManager.default.removeItem(at: file)
        }

        let url = try buildAvatarURL(credentials: credentials, username: username, size: size)
        let (data, response) = try await session.data(from: url)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode >= 200, httpResponse.statusCode < 300 else {
            throw ImageCacheError.downloadFailed(NSError(domain: "ImageCache", code: -1))
        }

        let contentType = httpResponse.value(forHTTPHeaderField: "Content-Type") ?? "image/jpeg"
        let ext = ImageCacheUtils.fileExtension(for: contentType)
        let fileName = "\(cacheId).\(ext)"
        let fileURL = directory.appendingPathComponent(fileName, isDirectory: false)

        try data.write(to: fileURL, options: [.atomic])

        let now = Int(Date().timeIntervalSince1970 * 1000)
        let record = CacheMetaRecord(
            key: "cover:\(username)",
            id: username,
            type: "cover",
            source: "explicit",
            triggersJson: nil,
            coverSize: size,
            sizeBytes: data.count,
            cachedAt: now,
            lastAccessedAt: now,
            removedFromServer: nil
        )

        let repo = CacheMetaRepository(db: db)
        try repo.upsert(record)

        return fileURL
    }

    func downloadCoverImage(coverArtId: String, size: String) async throws -> URL {
        guard let credentials = KeychainManager.retrieve() else {
            throw ImageCacheError.noCredentials
        }

        let directory = try ImageCacheUtils.cacheDirectoryURL(createIfNeeded: true)

        removeCoverImageFiles(cacheId: ImageCacheUtils.cacheId(for: coverArtId), in: directory)

        let url = try buildCoverArtURL(credentials: credentials, coverArtId: coverArtId, size: size)
        let (data, response) = try await session.data(from: url)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode >= 200, httpResponse.statusCode < 300 else {
            throw ImageCacheError.downloadFailed(NSError(domain: "ImageCache", code: -1))
        }

        let contentType = httpResponse.value(forHTTPHeaderField: "Content-Type") ?? "image/jpeg"
        let ext = ImageCacheUtils.fileExtension(for: contentType)
        let finalFileName = "\(ImageCacheUtils.cacheId(for: coverArtId)).\(ext)"
        let finalFileURL = directory.appendingPathComponent(finalFileName, isDirectory: false)

        try data.write(to: finalFileURL, options: [.atomic])

        let now = Int(Date().timeIntervalSince1970 * 1000)
        let record = CacheMetaRecord(
            key: "cover:\(coverArtId)",
            id: coverArtId,
            type: "cover",
            source: "explicit",
            triggersJson: nil,
            coverSize: size,
            sizeBytes: data.count,
            cachedAt: now,
            lastAccessedAt: now,
            removedFromServer: nil
        )

        let repo = CacheMetaRepository(db: db)
        try repo.upsert(record)
        notifyCoverImageCached(coverArtId: coverArtId)

        return finalFileURL
    }

    func storeCoverImage(coverArtId: String, data: Data, contentType: String, coverSize: String) throws -> URL {
        let directory = try ImageCacheUtils.cacheDirectoryURL(createIfNeeded: true)
        let ext = ImageCacheUtils.fileExtension(for: contentType)
        let fileName = "\(ImageCacheUtils.cacheId(for: coverArtId)).\(ext)"
        let fileURL = directory.appendingPathComponent(fileName, isDirectory: false)

        removeCoverImageFiles(cacheId: ImageCacheUtils.cacheId(for: coverArtId), in: directory)

        try data.write(to: fileURL, options: [.atomic])

        let now = Int(Date().timeIntervalSince1970 * 1000)
        let record = CacheMetaRecord(
            key: "cover:\(coverArtId)",
            id: coverArtId,
            type: "cover",
            source: "explicit",
            triggersJson: nil,
            coverSize: coverSize,
            sizeBytes: data.count,
            cachedAt: now,
            lastAccessedAt: now,
            removedFromServer: nil
        )

        let repo = CacheMetaRepository(db: db)
        try repo.upsert(record)
        notifyCoverImageCached(coverArtId: coverArtId)

        return fileURL
    }

    func resolveCoverImage(coverArtId: String) -> URL? {
        guard let directory = try? ImageCacheUtils.cacheDirectoryURL(createIfNeeded: false) else {
            return nil
        }

        let cacheId = ImageCacheUtils.cacheId(for: coverArtId)
        guard let fileURL = try? findCoverImageURL(cacheId: cacheId, in: directory) else {
            return nil
        }

        guard FileManager.default.fileExists(atPath: fileURL.path) else {
            return nil
        }

        // Update lastAccessedAt asynchronously (best effort)
        DispatchQueue.global(qos: .utility).async { [weak self] in
            guard let self else { return }
            let key = "cover:\(coverArtId)"
            let now = Int(Date().timeIntervalSince1970 * 1000)
            try? self.db.write { db in
                try CacheMetaRecord
                    .filter(Column("key") == key)
                    .updateAll(db, Column("lastAccessedAt").set(to: now))
            }
        }

        return fileURL
    }

    func deleteCoverImage(coverArtId: String) throws -> Bool {
        var deleted = false
        if let directory = try? ImageCacheUtils.cacheDirectoryURL(createIfNeeded: false),
           FileManager.default.fileExists(atPath: directory.path) {
            let cacheId = ImageCacheUtils.cacheId(for: coverArtId)
            deleted = removeCoverImageFiles(cacheId: cacheId, in: directory)
        }

        let key = "cover:\(coverArtId)"
        let repo = CacheMetaRepository(db: db)
        try repo.delete(key: key)

        return deleted
    }

    func clearCoverImages() throws -> Int {
        var deletedCount = 0
        if let directory = try? ImageCacheUtils.cacheDirectoryURL(createIfNeeded: false),
           FileManager.default.fileExists(atPath: directory.path) {
            let urls = try FileManager.default.contentsOfDirectory(
                at: directory,
                includingPropertiesForKeys: nil
            )

            for url in urls {
                try FileManager.default.removeItem(at: url)
                deletedCount += 1
            }
        }

        try db.write { db in
            try CacheMetaRecord
                .filter(Column("type") == "cover")
                .deleteAll(db)
        }

        return deletedCount
    }

    func getCoverImageSize(coverArtId: String) -> (sizeBytes: Int, coverSize: String?)? {
        guard let directory = try? ImageCacheUtils.cacheDirectoryURL(createIfNeeded: false),
              let fileURL = try? findCoverImageURL(
                  cacheId: ImageCacheUtils.cacheId(for: coverArtId),
                  in: directory
              ),
              FileManager.default.fileExists(atPath: fileURL.path),
              let attrs = try? FileManager.default.attributesOfItem(atPath: fileURL.path),
              let size = attrs[.size] as? NSNumber else {
            return nil
        }

        let key = "cover:\(coverArtId)"
        let record = try? CacheMetaRepository(db: db).getByKey(key)

        return (sizeBytes: size.intValue, coverSize: record?.coverSize)
    }

    // MARK: - Private Helpers

    private func notifyCoverImageCached(coverArtId: String) {
        NativeLogger.shared.info(
            "cover cached; posting notification coverArtId=\(coverArtId)",
            source: "ImageCache"
        )
        NotificationCenter.default.post(
            name: .aonsokuCoverImageCached,
            object: nil,
            userInfo: [ImageCacheNotification.coverArtIdKey: coverArtId]
        )
    }

    private func buildAvatarURL(credentials: ServerCredentials, username: String, size: String) throws -> URL {
        var params = SubsonicAuthBuilder.buildQueryParams(
            username: credentials.username,
            password: credentials.password,
            authType: credentials.authType,
            protocolVersion: credentials.protocolVersion
        )
        params["username"] = username
        params["size"] = size

        let baseString = "\(credentials.serverUrl)/rest/getAvatar"
        guard var components = URLComponents(string: baseString) else {
            throw ImageCacheError.invalidURL
        }
        components.queryItems = params.map { URLQueryItem(name: $0.key, value: $0.value) }

        guard let url = components.url else {
            throw ImageCacheError.invalidURL
        }
        return url
    }

    private func buildCoverArtURL(credentials: ServerCredentials, coverArtId: String, size: String) throws -> URL {
        var params = SubsonicAuthBuilder.buildQueryParams(
            username: credentials.username,
            password: credentials.password,
            authType: credentials.authType,
            protocolVersion: credentials.protocolVersion
        )
        params["id"] = coverArtId
        params["size"] = size

        let baseString = "\(credentials.serverUrl)/rest/getCoverArt"
        guard var components = URLComponents(string: baseString) else {
            throw ImageCacheError.invalidURL
        }
        components.queryItems = params.map { URLQueryItem(name: $0.key, value: $0.value) }

        guard let url = components.url else {
            throw ImageCacheError.invalidURL
        }
        return url
    }

    private func findCoverImageURL(cacheId: String, in directory: URL) throws -> URL? {
        ImageCacheUtils.cachedImageExtensions
            .map { directory.appendingPathComponent("\(cacheId).\($0)", isDirectory: false) }
            .first { FileManager.default.fileExists(atPath: $0.path) }
    }

    @discardableResult
    private func removeCoverImageFiles(cacheId: String, in directory: URL) -> Bool {
        var deleted = false
        for ext in ImageCacheUtils.cachedImageExtensions {
            let url = directory.appendingPathComponent("\(cacheId).\(ext)", isDirectory: false)
            guard FileManager.default.fileExists(atPath: url.path) else {
                continue
            }
            do {
                try FileManager.default.removeItem(at: url)
                deleted = true
            } catch {}
        }
        return deleted
    }
}

File: capacitor-plugins/capacitor-native/android/src/main/java/github/realtvop/aonsoku/plugins/data/image/ImageCacheManager.kt
package github.realtvop.aonsoku.plugins.data.image

import github.realtvop.aonsoku.plugins.bridge.ServerCredentials
import github.realtvop.aonsoku.plugins.bridge.SubsonicAuthBuilder
import github.realtvop.aonsoku.plugins.data.db.dao.CacheMetaDao
import github.realtvop.aonsoku.plugins.data.db.entity.CacheMetaEntity
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.util.concurrent.TimeUnit

class ImageCacheManager(private val cacheDir: File, private val cacheMetaDao: CacheMetaDao) {
    private val client = OkHttpClient.Builder().connectTimeout(30, TimeUnit.SECONDS).readTimeout(120, TimeUnit.SECONDS).build()

    suspend fun downloadCoverImage(coverArtId: String, size: String, credentials: ServerCredentials): File {
        val dir = ImageCacheUtils.cacheDirectory(cacheDir, true); val cid = ImageCacheUtils.cacheId(coverArtId)
        ImageCacheUtils.deleteCoverImageFiles(dir, cid)
        val p = SubsonicAuthBuilder.buildQueryParams(credentials.username, credentials.password, credentials.authType, credentials.protocolVersion).toMutableMap().apply { put("id", coverArtId); put("size", size) }
        val url = "${credentials.serverUrl.trimEnd('/')}/rest/getCoverArt?${p.entries.joinToString("&") { (k, v) -> "${java.net.URLEncoder.encode(k, "UTF-8")}=${java.net.URLEncoder.encode(v, "UTF-8")}" }}"
        val resp = client.newCall(Request.Builder().url(url).get().build()).execute()
        if (!resp.isSuccessful) throw ImageCacheError.DownloadFailed(Exception("HTTP ${resp.code}"))
        val ct = resp.header("Content-Type", "image/jpeg") ?: "image/jpeg"; val ext = ImageCacheUtils.fileExtension(ct)
        val data = resp.body?.bytes() ?: throw ImageCacheError.DownloadFailed(Exception("Empty body"))
        val file = File(dir, "$cid.$ext"); file.writeBytes(data)
        val now = System.currentTimeMillis()
        cacheMetaDao.upsert(CacheMetaEntity("cover:$coverArtId", coverArtId, "cover", "explicit", coverSize = size, sizeBytes = data.size.toLong(), cachedAt = now, lastAccessedAt = now))
        return file
    }

    suspend fun downloadAvatar(username: String, size: String, credentials: ServerCredentials): File {
        val dir = ImageCacheUtils.cacheDirectory(cacheDir, true); val cid = ImageCacheUtils.cacheId(username)
        ImageCacheUtils.deleteCoverImageFiles(dir, cid)
        val p = SubsonicAuthBuilder.buildQueryParams(credentials.username, credentials.password, credentials.authType, credentials.protocolVersion).toMutableMap().apply { put("username", username); put("size", size) }
        val url = "${credentials.serverUrl.trimEnd('/')}/rest/getAvatar?${p.entries.joinToString("&") { (k, v) -> "${java.net.URLEncoder.encode(k, "UTF-8")}=${java.net.URLEncoder.encode(v, "UTF-8")}" }}"
        val resp = client.newCall(Request.Builder().url(url).get().build()).execute()
        if (!resp.isSuccessful) throw ImageCacheError.DownloadFailed(Exception("HTTP ${resp.code}"))
        val ct = resp.header("Content-Type", "image/jpeg") ?: "image/jpeg"; val ext = ImageCacheUtils.fileExtension(ct)
        val data = resp.body?.bytes() ?: throw ImageCacheError.DownloadFailed(Exception("Empty body"))
        val file = File(dir, "$cid.$ext"); file.writeBytes(data)
        val now = System.currentTimeMillis()
        cacheMetaDao.upsert(CacheMetaEntity("cover:$username", username, "cover", "explicit", coverSize = size, sizeBytes = data.size.toLong(), cachedAt = now, lastAccessedAt = now))
        return file
    }

    suspend fun storeCoverImage(coverArtId: String, data: ByteArray, contentType: String, coverSize: String): File {
        val dir = ImageCacheUtils.cacheDirectory(cacheDir, true); val cid = ImageCacheUtils.cacheId(coverArtId); val ext = ImageCacheUtils.fileExtension(contentType)
        val file = File(dir, "$cid.$ext"); ImageCacheUtils.deleteCoverImageFiles(dir, cid); file.writeBytes(data)
        val now = System.currentTimeMillis()
        cacheMetaDao.upsert(CacheMetaEntity("cover:$coverArtId", coverArtId, "cover", "explicit", coverSize = coverSize, sizeBytes = data.size.toLong(), cachedAt = now, lastAccessedAt = now))
        return file
    }

    suspend fun resolveCoverImage(coverArtId: String): File? {
        val dir = ImageCacheUtils.cacheDirectory(cacheDir, false); if (!dir.exists()) return null
        val cid = ImageCacheUtils.cacheId(coverArtId); val f = ImageCacheUtils.coverImageFile(dir, cid) ?: return null
        if (!f.exists()) return null
        try { cacheMetaDao.getByKey("cover:$coverArtId")?.let { cacheMetaDao.upsert(it.copy(lastAccessedAt = System.currentTimeMillis())) } } catch (_: Exception) {}
        return f
    }

    suspend fun deleteCoverImage(coverArtId: String): Boolean {
        val dir = ImageCacheUtils.cacheDirectory(cacheDir, false); if (!dir.exists()) return false; val cid = ImageCacheUtils.cacheId(coverArtId)
        val d = ImageCacheUtils.deleteCoverImageFiles(dir, cid)
        try { cacheMetaDao.delete("cover:$coverArtId") } catch (_: Exception) {}; return d
    }

    suspend fun clearCoverImages(): Int {
        val dir = ImageCacheUtils.cacheDirectory(cacheDir, false); if (!dir.exists()) return 0
        val files = dir.listFiles() ?: return 0; files.forEach { it.delete() }
        try { cacheMetaDao.deleteByType("cover") } catch (_: Exception) {}; return files.size
    }

    suspend fun getCoverImageSize(coverArtId: String): Pair<Long, String?>? {
        val dir = ImageCacheUtils.cacheDirectory(cacheDir, false); val cid = ImageCacheUtils.cacheId(coverArtId)
        val f = ImageCacheUtils.coverImageFile(dir, cid) ?: return null; if (!f.exists()) return null
        val r = try { cacheMetaDao.getByKey("cover:$coverArtId") } catch (_: Exception) { null }
        return Pair(f.length(), r?.coverSize)
    }
}
